### PR TITLE
Support go plugins for forges and agent backends

### PR DESCRIPTION
**Pull request by qwerty287**
As of #2520 

Support to load new forges and agent backends at runtime using go's plugin system. (https://pkg.go.dev/plugin)

I also added a simple example addon (a new forge which just prints log statements), it should be removed later of course, but you can see an example.

### CODE REVIEW

**Comment by anbraten** [comment]
Looks pretty nice already. 
Two questions / ideas that came up to me:
- How can we add a plugin to the docker image? Would we create a docker image using the "normal" image as base and simply add the plugin binary?
- Can we combine this with the requested compiler, secrets, registry addons somehow? Problem would be as those should be selectable by users they would need to be executed in a rather separated environment like a step by the agent and therefore we might need to consider sth like <https://github.com/hashicorp/go-plugin>

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+```sh
+go build -buildmode plugin
+```
+
+The output file is your addon which is now ready to use.
+
+## Restrictions
+
+Addons must directly directly depend on Woodpecker's core (`go.woodpecker-ci.org/woodpecker/woodpecker`).
```
Comment (mzampetakis) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:40]: Nitpick: Duplicate `directly`.

**Comment by mzampetakis** [comment]
This looks great! 

It seems that any loaded forge through this add-on system will override the server's configuration for any forge. This could be documented at someplace (e.g. `docs/docs/30-administration/75-addons/00-addons.md`) so that the administrators will have a more clear view about server's bevahiour.

**Comment by mzampetakis** [comment]
Is there any particular reason for using `hashicorp/go-plugin` instead of the built in one? 
I believe that it might be more convenient to stick with the go's built in one in terms of upgradability.
What do you think?

**Comment by anbraten** [comment]
"users" would be all users (including non-instance-admins) in my idea. The idea I had some time ago was that the server asks an agent to start a service (exp container) (similar to running a pipeline somewhere). This service would then provide an http endpoint. This endpoint could for example provide the config-service endpoint, so it could be used as an compiler. So as soon as the server needs the config it would check that the service is running and forwards the http request through the grpc connection via the agent to the service. But all of that was just a first idea.

**Comment by mzampetakis** [comment]
Is this completed or blocked by something? I could run some tests with a dummy addon forge to check if everything is working as expected. Please let me know if I could provide any assistance.

**Comment by anbraten** [comment]
> Your suggestion won't work. hashicorp/go-plugin does not support RPC over remote networks, only locally.

That was just a reference. I think we need to create our own http endpoints definitions if we want to go that path. I would use an http api similar to what we have for the config-service already as grpc wouldn't be resistent to possible changes.

**Comment by anbraten** [comment]
> So this would be totally different from this approach. Can I revert it back to go plugins?

Sure, I would be fine adding the go-plugin approach for now as experimental and would not document it, so we could drop / replace it at any point.

**Comment by mzampetakis** [comment]
> > So this would be totally different from this approach. Can I revert it back to go plugins?
> 
> Sure, I would be fine adding the go-plugin approach for now as experimental and would not document it, so we could drop / replace it at any point.

I am fond of moving towards the native go plugin approach as wel. 

Being not documented/experimental means that it could be dropped at some point of time right? So, any implemented addons that will use this approach to integrate with woodpecker will be dropped as well right?

If that's the case I would recommend to be documented but noted as experimental (alongside the possibility of dropping this in the future) in order to be crystal clear but also be able to collect any feedback from implemented addons.

File: cmd/server/flags.go [hunk 2]
```diff
@@ -256,6 +256,10 @@ var flags = append([]cli.Flag{
 		Name:    "enable-swagger",
 		Value:   true,
 	},
+	&cli.StringSliceFlag{
+		EnvVars: []string{"WOODPECKER_ADDONS"},
+		Name:    "addons",
```
Comment (mzampetakis) [inline cmd/server/flags.go:261]: Maybe the `usage` could be added here.

File: docs/docs/30-administration/75-addons/00-overview.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+```diff
+# docker-compose.yml
+version: '3'
+
+services:
+  woodpecker-server:
+    [...]
+    environment:
++     - WOODPECKER_PLUGIN=/path/to/your/addon/file.so
```
Comment (mzampetakis) [inline docs/docs/30-administration/75-addons/00-overview.md:33]: I think it's `WOODPECKER_ADDONS` here:
```suggestion
+     - WOODPECKER_ADDONS=/path/to/your/addon/file.so
```

File: docs/docs/30-administration/75-addons/00-overview.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+You may need to [mount the addon file as volume](https://docs.docker.com/storage/volumes/#create-and-manage-volumes) to access it from inside the Docker container.
+
+You can list multiple addons, Woodpecker will automatically determine their type. If you specify multiple addons with the same type, only the first one will be used.
```
Comment (mzampetakis) [inline docs/docs/30-administration/75-addons/00-overview.md:38]: Maybe it should state that any forge addon will override other setup forges.

File: cmd/agent/agent.go [hunk 5]
```diff
@@ -247,6 +251,25 @@ func run(c *cli.Context) error {
 	return nil
 }
 
+func getEngine(backendCtx context.Context, engineName string, addons []string) (types.Engine, error) {
+	addonEngine, err := addon.Load[types.Engine](addons, addonTypes.TypeEngine)
+	if err != nil {
+		log.Error().Err(err).Msg("cannot load addon")
```
Comment (anbraten) [inline cmd/agent/agent.go:257]:
```suggestion
		log.Error().Err(err).Msg("cannot load backend addon")
```

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+   2. An error. If this error is not `nil`, Woodpecker exits.
+
+Directly import Woodpecker's Go package (`go.woodpecker-ci.org/woodpecker/woodpecker`) and use the interfaces and types defined there.
+
+### Return types
+
+| Addon type | Return type                                                                  |
+| ---------- | ---------------------------------------------------------------------------- |
+| `Forge`    | `"go.woodpecker-ci.org/woodpecker/woodpecker/server/forge".Forge`            |
+| `Engine`   | `"go.woodpecker-ci.org/woodpecker/woodpecker/pipeline/backend/types".Engine` |
```
Comment (anbraten) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:28]:
```suggestion
| `Forge`    | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/forge".Forge`            |
| `Engine`   | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/pipeline/backend/types".Engine` |
```

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+In addition to this, addons are only supported on Linux, FreeBSD and macOS.
+
+:::info
+It is recommended to at least support the latest released version of Woodpecker.
+:::
+
+### Compile for different versions
+
+As long as there were no changes to Woodpecker's interfaces or they are backwards-compatible, you can easily compile the addon for multiple version by changing the version of `go.woodpecker-ci.org/woodpecker/woodpecker` using `go get` before compiling.
```
Comment (anbraten) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:54]:
```suggestion
As long as there were no changes to Woodpecker's interfaces or they are backwards-compatible, you can easily compile the addon for multiple version by changing the version of `go.woodpecker-ci.org/woodpecker/woodpecker/v2` using `go get` before compiling.
```

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+import (
+  "context"
+  "net/http"
+
+  "github.com/rs/zerolog"
+  "go.woodpecker-ci.org/woodpecker/woodpecker/server/forge"
+  forge_types "go.woodpecker-ci.org/woodpecker/woodpecker/server/forge/types"
+  "go.woodpecker-ci.org/woodpecker/woodpecker/server/model"
+  addon_types "go.woodpecker-ci.org/woodpecker/woodpecker/shared/addon/types"
```
Comment (anbraten) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:73]:
```suggestion
  "go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/forge"
  forge_types "go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/forge/types"
  "go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/model"
  addon_types "go.woodpecker-ci.org/woodpecker/woodpecker/v2/shared/addon/types"
```

File: shared/addon/addon.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+			}
+			pluginCache[file] = p
+		}
+
+		typeLookup, err := pluginCache[file].Lookup("Type")
+		if err != nil {
+			return nil, err
+		}
+		if addonType, is := typeLookup.(*types.Type); !is {
+			return nil, errors.New("addon type has incorrect type")
```
Comment (anbraten) [inline shared/addon/addon.go:38]:
```suggestion
			return nil, errors.New("addon type is incorrect")
```

File: shared/addon/addon.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		} else if *addonType != t {
+			continue
+		}
+
+		mainLookup, err := pluginCache[file].Lookup("Addon")
+		if err != nil {
+			return nil, err
+		}
+		fmt.Println(reflect.TypeOf(mainLookup))
```
Comment (anbraten) [inline shared/addon/addon.go:47]:
```suggestion
```

File: shared/addon/addon.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+		main, is := mainLookup.(func(zerolog.Logger, []string) (T, error))
+		if !is {
+			return nil, errors.New("addon main has incorrect type")
```
Comment (anbraten) [inline shared/addon/addon.go:50]:
```suggestion
			return nil, errors.New("addon main function has incorrect type")
```

File: docs/docs/30-administration/75-addons/00-overview.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    [...]
+    environment:
++     - WOODPECKER_ADDONS=/path/to/your/addon/file.so
+```
+
+You may need to [mount the addon file as volume](https://docs.docker.com/storage/volumes/#create-and-manage-volumes) to access it from inside the Docker container.
+
+You can list multiple addons, Woodpecker will automatically determine their type. If you specify multiple addons with the same type, only the first one will be used.
+
+Using an addon always overwrites Woodpecker's internal setup. This means, that a forge addon will be used if specified, no matter what's configured for the forges natively supported by Woodpecker.
```
[the remark this answers is not in the document]
Comment (anbraten) [reply docs/docs/30-administration/75-addons/00-overview.md:44]: Yep, just wanted to mention that.

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+### Return types
+
+| Addon type | Return type                                                                     |
+| ---------- | ------------------------------------------------------------------------------- |
+| `Forge`    | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/forge".Forge`            |
+| `Engine`   | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/pipeline/backend/types".Engine` |
+
+## Compiling
+
+After you wrote your addon code, compile your addon:
```
Comment (mzampetakis) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:32]:
```suggestion
After you write your addon code, compile your addon:
```

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+In addition to this, addons are only supported on Linux, FreeBSD and macOS.
+
+:::info
+It is recommended to at least support the latest released version of Woodpecker.
+:::
+
+### Compile for different versions
+
+As long as there were no changes to Woodpecker's interfaces or they are backwards-compatible, you can easily compile the addon for multiple version by changing the version of `go.woodpecker-ci.org/woodpecker/woodpecker/v2` using `go get` before compiling.
```
Comment (mzampetakis) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:54]:
```suggestion
As long as there are no changes to Woodpecker's interfaces or they are backwards-compatible, you can easily compile the addon for multiple version by changing the version of `go.woodpecker-ci.org/woodpecker/woodpecker/v2` using `go get` before compiling.
```

File: cmd/agent/flags.go [hunk 15]
```diff
@@ -97,4 +97,8 @@ var flags = []cli.Flag{
 		Usage:   "backend to run pipelines on",
 		Value:   "auto-detect",
 	},
+	&cli.StringSliceFlag{
+		EnvVars: []string{"WOODPECKER_ADDONS"},
+		Name:    "addons",
```
Comment (6543) [inline cmd/agent/flags.go:102]:
```suggestion
		Name:    "addons",
		Usage: "list of addon files",
```

**Review by mzampetakis — APPROVED** [review]
Looks great! ♥️

File: docs/docs/30-administration/75-addons/20-creating-addons.md [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+   2. An error. If this error is not `nil`, Woodpecker exits.
+
+Directly import Woodpecker's Go package (`go.woodpecker-ci.org/woodpecker/woodpecker/v2`) and use the interfaces and types defined there.
+
+### Return types
+
+| Addon type | Return type                                                                     |
+| ---------- | ------------------------------------------------------------------------------- |
+| `Forge`    | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/server/forge".Forge`            |
+| `Engine`   | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/pipeline/backend/types".Engine` |
```
Comment (6543) [inline docs/docs/30-administration/75-addons/20-creating-addons.md:28]:
```suggestion
| `Backend`   | `"go.woodpecker-ci.org/woodpecker/woodpecker/v2/pipeline/backend/types".Backend` |
```

**Review by 6543 — APPROVED** [review]

**Pull request merged**